diff --git a/android/gki_aarch64_modules b/android/gki_aarch64_modules
index 3d8888c..e69de29 100644
--- a/android/gki_aarch64_modules
+++ b/android/gki_aarch64_modules
@@ -1,2 +0,0 @@
-mm/zsmalloc.ko
-drivers/block/zram/zram.ko
diff --git a/android/gki_system_dlkm_modules b/android/gki_system_dlkm_modules
index 83e76f5..e69de29 100644
--- a/android/gki_system_dlkm_modules
+++ b/android/gki_system_dlkm_modules
@@ -1,2 +0,0 @@
-drivers/block/zram/zram.ko
-mm/zsmalloc.ko
