diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 563e62b6a..14901d358 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -11474,7 +11474,7 @@ struct sk_buff *__cfg80211_alloc_event_skb(struct wiphy *wiphy,
 	return __cfg80211_alloc_vendor_skb(rdev, wdev, approxlen, portid, 0,
 					   cmd, attr, info, gfp);
 }
-EXPORT_SYMBOL(__cfg80211_alloc_event_skb);
+
 
 void __cfg80211_send_event_skb(struct sk_buff *skb, gfp_t gfp)
 {
