diff --git a/arch/Kconfig b/arch/Kconfig
index 97e62dcb2..88b4eb0a0 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -253,6 +253,9 @@ config ARCH_HAS_KEEPINITRD
 config ARCH_HAS_SET_MEMORY
 	bool
 
+config ARCH_HAS_HW_PTE_YOUNG
+	bool
+
 # Select if arch has all set_direct_map_invalid/default() functions
 config ARCH_HAS_SET_DIRECT_MAP
 	bool
diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 9b4a1830d..3eecdd6f1 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -34,6 +34,7 @@ config ARM64
 	select ARCH_HAS_NON_OVERLAPPING_ADDRESS_SPACE
 	select ARCH_HAS_PTE_DEVMAP
 	select ARCH_HAS_PTE_SPECIAL
+	select ARCH_HAS_HW_PTE_YOUNG
 	select ARCH_HAS_SETUP_DMA_OPS
 	select ARCH_HAS_SET_DIRECT_MAP
 	select ARCH_HAS_SET_MEMORY
diff --git a/arch/arm64/include/asm/pgtable.h b/arch/arm64/include/asm/pgtable.h
index 46b1fda88..f6e21c066 100644
--- a/arch/arm64/include/asm/pgtable.h
+++ b/arch/arm64/include/asm/pgtable.h
@@ -1016,7 +1016,7 @@ static inline void update_mmu_cache(struct vm_area_struct *vma,
  * page after fork() + CoW for pfn mappings. We don't always have a
  * hardware-managed access flag on arm64.
  */
-#define arch_has_hw_pte_young		cpu_has_hw_af
+#define arch_has_hw_pte_young()		(true) /* Seiran: force HW AF for MGLRU MM walk on SM6225 */
 
 /*
  * Experimentally, it's cheap to set the access flag in hardware and we
diff --git a/mm/Kconfig b/mm/Kconfig
index fb3cfd168..061bc0228 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -944,6 +944,10 @@ config LRU_GEN_STATS
 	  from evicted generations for debugging purpose.
 
 	  This option has a per-memcg and per-node memory overhead.
+
+config LRU_GEN_WALKS_MMU
+	def_bool y
+	depends on LRU_GEN && ARCH_HAS_HW_PTE_YOUNG
 # }
 
 source "mm/damon/Kconfig"
