diff --git a/kernel/time/posix-cpu-timers.c b/kernel/time/posix-cpu-timers.c
index 88e781772..f3664539e 100644
--- a/kernel/time/posix-cpu-timers.c
+++ b/kernel/time/posix-cpu-timers.c
@@ -519,7 +519,7 @@ static struct task_struct *timer_lock_sighand(struct k_itimer *timer, unsigned l
 	enum pid_type type = clock_pid_type(timer->it_clock);
 	struct cpu_timer *ctmr = &timer->it.cpu;
 
-	guard(rcu)();
+	rcu_read_lock();
 
 	for (;;) {
 		struct task_struct *t = pid_task(timer->it.cpu.pid, type);
@@ -529,8 +529,10 @@ static struct task_struct *timer_lock_sighand(struct k_itimer *timer, unsigned l
 			break;
 
 		/* Try to lock the task's sighand */
-		if (lock_task_sighand(t, flags))
+		if (lock_task_sighand(t, flags)) {
+			rcu_read_unlock();
 			return t;
+		}
 
 		/*
 		 * The next PID lookup might either fail or return the new
@@ -556,6 +558,7 @@ static struct task_struct *timer_lock_sighand(struct k_itimer *timer, unsigned l
 	 */
 	smp_rmb();
 	WARN_ON_ONCE(ctmr->head || timerqueue_node_queued(&ctmr->node));
+	rcu_read_unlock();
 	return NULL;
 }
 
