From e6036fddf42258ac02c7ed65362ec4ce196e6872 Mon Sep 17 00:00:00 2001
From: chickendrop89 <chickendrop89@gmail.com>
Date: Wed, 30 Jul 2025 23:12:33 +0200
Subject: [PATCH] kernel: printk: add a missing comma

* Oops

Signed-off-by: chickendrop89 <chickendrop89@gmail.com>
---
 kernel/printk/printk.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/printk/printk.c b/kernel/printk/printk.c
index 78a0fe68227f..fa9472cd47a5 100644
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -2159,7 +2159,7 @@ __maybe_unused static bool should_filter_vendor_message(const char *text)
         #endif
             "st21n", /* st21nfc  */
             "[Awin", /* [Awinic] */
-			"qpnp_"  /* qpnp_vib_brightness_set */
+			"qpnp_", /* qpnp_vib_brightness_set */
             NULL
     };
 
